include/grpc/event_engine/memory_allocator.h

```diff
@@ -0,0 +1,242 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#ifndef GRPC_EVENT_ENGINE_MEMORY_ALLOCATOR_H
+#define GRPC_EVENT_ENGINE_MEMORY_ALLOCATOR_H
+
+#include <grpc/impl/codegen/port_platform.h>
+
+#include <algorithm>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+#include <grpc/slice.h>
+
+// forward-declaring an internal struct, not used publicly.
+struct grpc_slice_buffer;
+
+namespace grpc_event_engine {
+namespace experimental {
+
+// TODO(nnoble): needs implementation
+class SliceBuffer {
+ public:
+  SliceBuffer() { abort(); }
+  explicit SliceBuffer(grpc_slice_buffer*) { abort(); }
+
+  grpc_slice_buffer* RawSliceBuffer() { return slice_buffer_; }
+
+ private:
+  grpc_slice_buffer* slice_buffer_;
+};
+
+// Reservation request - how much memory do we want to allocate?
+class MemoryRequest {
+ public:
+  // Request a fixed amount of memory.
+  // NOLINTNEXTLINE(google-explicit-constructor)
+  MemoryRequest(size_t n) : min_(n), max_(n) {}
+  // Request a range of memory.
+  MemoryRequest(size_t min, size_t max) : min_(std::min(min, max)), max_(max) {}
+
+  // Increase the size by amount
+  MemoryRequest Increase(size_t amount) const {
+    return MemoryRequest(min_ + amount, max_ + amount);
+  }
+
+  size_t min() const { return min_; }
+  size_t max() const { return max_; }
+
+ private:
+  size_t min_;
+  size_t max_;
+};
+
+class BasicMemoryAllocator
+    : public std::enable_shared_from_this<BasicMemoryAllocator> {
+ public:
+  BasicMemoryAllocator() {}
+  virtual ~BasicMemoryAllocator() {}
+
+  BasicMemoryAllocator(const BasicMemoryAllocator&) = delete;
+  BasicMemoryAllocator& operator=(const BasicMemoryAllocator&) = delete;
+
+  // Reserve bytes from the quota.
+  // If we enter overcommit, reclamation will begin concurrently.
+  // Returns the number of bytes reserved.
+  virtual size_t Reserve(MemoryRequest request) = 0;
+
+  // Release some bytes that were previously reserved.
+  virtual void Release(size_t n) = 0;
+
+  // Shutdown this allocator.
+  // Further usage of Reserve() is undefined behavior.
+  virtual void Shutdown() = 0;
+};
+
+class MemoryAllocator {
+ public:
+  // Construct a MemoryAllocator given a BasicMemoryAllocator implementation.
+  // The constructed MemoryAllocator will call BasicMemoryAllocator::Shutdown()
+  // upon destruction.
+  explicit MemoryAllocator(std::shared_ptr<BasicMemoryAllocator> allocator)
+      : allocator_(allocator) {}
+  ~MemoryAllocator() {
+    if (allocator_ != nullptr) allocator_->Shutdown();
```

Comment:
Explicitly shutting down a `std::shared_ptr` is unfriendly, no? Why not make it a unique_ptr if this class controls the BaseMemoryAllocator's usable lifetime?